core/Network/TLS/Handshake/Client.hs

```diff
@@ -415,3 +429,17 @@ processCertificateRequest ctx p = processServerHelloDone ctx p
 processServerHelloDone :: Context -> Handshake -> IO (RecvState m)
 processServerHelloDone _ ServerHelloDone = return RecvStateDone
 processServerHelloDone _ p = unexpected (show p) (Just "server hello data")
+
+requiredCertKeyUsage :: Cipher -> Maybe ExtKeyUsageFlag
+requiredCertKeyUsage cipher =
+    case cipherKeyExchange cipher of
+        CipherKeyExchange_RSA         -> Just KeyUsage_keyEncipherment
+        CipherKeyExchange_DH_Anon     -> Nothing
+        CipherKeyExchange_DHE_RSA     -> Just KeyUsage_digitalSignature
+        CipherKeyExchange_ECDHE_RSA   -> Just KeyUsage_digitalSignature
+        CipherKeyExchange_DHE_DSS     -> Just KeyUsage_digitalSignature
+        CipherKeyExchange_DH_DSS      -> Just KeyUsage_keyAgreement
+        CipherKeyExchange_DH_RSA      -> Just KeyUsage_keyAgreement
+        CipherKeyExchange_ECDH_ECDSA  -> Just KeyUsage_keyAgreement
+        CipherKeyExchange_ECDH_RSA    -> Just KeyUsage_keyAgreement
+        CipherKeyExchange_ECDHE_ECDSA -> Just KeyUsage_digitalSignature
```

Comment:
I should further note that with the only mention of `keyUsage` enforcement in TLS RFCs being in TLS 1.3 (RFC8446, appendix E.8) it is rather ironic that we're enforcing the constraint based on the selected cipher, since in TLS 1.3 the ciphers don't actually select the key exchange method, that's based on the signature algorithms extension and certificate type. So the check is entirely ineffective with TLS 1.3!

So perhaps this PR not only causes some interoperability issues, but also fails to achieve the goals it set out to meet. :-(